lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp

```diff
@@ -70,164 +32,325 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
+/// beginning of shared memory object.
+Value computeSwizzledOffset(ConversionPatternRewriter &rewriter, Location loc,
+                            const Indexes &i, const DimNumbers &dim,
+                            Value bTileOffset, Value nonKTileOffset,
+                            unsigned shapePerCTABTile,
+                            unsigned shapePerCTANonKTile,
+                            SharedEncodingAttr sharedLayout,
+                            ArrayRef<int64_t> opTensorShape,
+                            ArrayRef<Value> strides) {
+  Value offset = i32_val(0);
+  // Compute unswizzled multi dim coordinates in shared memmory object
+  SmallVector<Value> elemMultiDimIndices(3);
+  elemMultiDimIndices[dim.batch] =
+      add(bTileOffset, i32_val(i.bTile * shapePerCTABTile + i.b));
+  elemMultiDimIndices[dim.nonK] =
+      add(nonKTileOffset, i32_val(i.nonKTile * shapePerCTANonKTile + i.nonK));
+  elemMultiDimIndices[dim.k] = i32_val(i.k);
+
+  // Apply swizzling pattern to fastest dimension
+  SmallVector<Value> swizzledIndices =
+      swizzleIndices(rewriter, loc, elemMultiDimIndices, sharedLayout);
+
+  // Linearize shared mem object dimensions into flat offset
+  for (int d = 0; d < 3; ++d) {
+    // wrap index if it is larger than tensor
+    auto wrappedDimIndex = urem(swizzledIndices[d], i32_val(opTensorShape[d]));
+    auto dimOffset = mul(wrappedDimIndex, strides[d]);
+    offset = add(offset, dimOffset);
+  }
+  return offset;
+}
 
-  return getStructFromValueTable(vas, rewriter, loc, typeConverter, elemTy);
+/// Computes memory offset of a given element relative to the
+/// first element loaded by a thread.
+Value computeNonSwizzledOffset(ConversionPatternRewriter &rewriter,
+                               Location loc, const Indexes &i,
+                               const DimNumbers &dim,
+                               ArrayRef<int64_t> tensorShape,
+                               unsigned shapePerCTABTile,
+                               unsigned shapePerCTANonKTile,
+                               ArrayRef<Value> strides) {
+  SmallVector<Value> offsetIndices(3);
+  offsetIndices[dim.batch] =
+      i32_val((i.bTile * shapePerCTABTile + i.b) % tensorShape[dim.batch]);
+  offsetIndices[dim.nonK] = i32_val(
+      (i.nonKTile * shapePerCTANonKTile + i.nonK) % tensorShape[dim.nonK]);
+  offsetIndices[dim.k] = i32_val(i.k);
+
+  Value offset = i32_val(0);
+  for (int d = 0; d < 3; ++d)
+    offset = add(offset, mul(offsetIndices[d], strides[d]));
+  return offset;
 }
 
-Value loadBFMA(Value B, Value llB, BlockedEncodingAttr dLayout, Value thread,
-               Location loc, const LLVMTypeConverter *typeConverter,
-               ConversionPatternRewriter &rewriter) {
-  auto bTensorTy = cast<MemDescType>(B.getType());
-  auto bLayout = cast<SharedEncodingAttr>(bTensorTy.getEncoding());
-  auto bShapePerCTA = getShapePerCTA(bTensorTy);
+Value loadFMAOp(Value dotOp, Value llA, BlockedEncodingAttr dLayout,
```

Comment:
The 1st arg is the src of local_local. Can you change it to something like "srcVal"?